src/Orleans.Runtime/Catalog/ActivationData.cs

```diff
@@ -586,8 +586,13 @@ public void ResetKeepAliveRequest()
                     OnInactive = new List<Action>();
                 }
                 OnInactive.Add(action);
+                if (!IsCurrentlyExecuting)
+                {
+                    RunOnInactive();
```

Comment:
Locks are reentrant, so it's ok, but the lock will be re-acquired by this call. Unsure what the cost of that is, if at all significant. An alternative would be an "*UnderLock" or "Unsafe*" alternative without a lock, to signify that the lock has already been taken.